### PR TITLE
update mysql-compatibility.md & add auto-increment.md

### PR DESCRIPTION
**Pull request by tangenta**
### What is changed, added or deleted? (Required)
- Update the `auto_increment` part and DDL part of mysql-compatibility.md.
- Add document for `auto_increment`.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### CODE REVIEW

File: auto-increment.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+TiDB 实现 `AUTO_INCREMENT` 隐式分配的原理是，对于每一个自增列，都使用一个全局可见的键值对用于记录当前已分配的最大 ID。由于分布式环境下的节点通信存在一定开销，为了避免写请求放大的问题，每个 TiDB 节点在分配 ID 时，都申请一段 ID 作为缓存，用完之后再去取下一段，而不是每次分配都向存储节点申请。例如，对于以下新建的表：
+
+```sql
+create table t(id int unique key AUTO_INCREMENT, c int);
+```
+
+假设集群中有两个 TiDB 实例 A 和 B，如果向 A 和 B 分别对 `t` 执行一条插入语句：
+
+```sql
+insert into t (c) (1)
```
Comment (AilinKid) [inline auto-increment.md:80]:
```suggestion
insert into t (c) values (1);
```

File: auto-increment.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+```
+
+实例 A 可能会缓存 [1,30000] 的自增 ID，而实例 B 则可能缓存 [30001,60000] 的自增 ID。这些缓存的 ID 将被作为缺省值，随着执行将来的插入语句，顺序地填充到 `AUTO_INCREMENT` 列中。
```
Comment (AilinKid) [inline auto-increment.md:83]:
```suggestion
实例 A 可能会缓存 [1,30000] 的自增 ID，而实例 B 则可能缓存 [30001,60000] 的自增 ID。各自节点缓存的 ID 将随着执行将来插入语句被作为缺省值，顺序地填充到 `AUTO_INCREMENT` 列中。
```

File: auto-increment.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+create table t (a int primary key AUTO_INCREMENT)
+```
+
+```sql
+insert into t values (), (), (), ()
+```
+
+即使存在正在执行并发写入的其他 TiDB 实例，或者当前实例剩余的缓存 ID 数量不够，都不会影响分配值的连续性。
+
+### 和 _tidb_rowid 的关联
```
Comment (AilinKid) [inline auto-increment.md:120]:
```suggestion
### _tidb_rowid 的关联性
```

File: auto-increment.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+从 v3.0.9 和 v4.0.rc.1 开始，和 MySQL 的行为类似，自增列隐式分配的值遵循 session 变量 `@@auto_increment_increment` 和 `@@auto_increment_offset` 的控制，其中自增列隐式分配的值(ID)将满足式子 `(ID - auto_increment_offset) % auto_increment_increment == 0`。
+
+## 使用限制
+
+目前在 TiDB 中使用 `AUTO_INCREMENT` 有以下限制：
+
+- 必须定义在主键或者唯一索引的列上。
+- 只能定义在类型为整数、`FLOAT` 或 `DOUBLE` 的列上
+- 不支持与列的默认值 `DEFAULT` 同时指定在同一列上。。
```
Comment (AilinKid) [inline auto-increment.md:159]: 这个句号应该放到上面

File: mysql-compatibility.md [hunk 5]
```diff
@@ -104,34 +61,23 @@ TiDB 支持常用的 MySQL 内建函数，但是不是所有的函数都已经
...
-    - `PARTITION BY LIST`
-    - `PARTITION BY KEY`
-    - `SUBPARTITION`
-    - `{CHECK|EXCHANGE|TRUNCATE|OPTIMIZE|REPAIR|IMPORT|DISCARD|REBUILD|REORGANIZE} PARTITION`
+在 TiDB 中，运行的 DDL 操作不会影响对表的读取或写入。但是，目前 DDL 以下变更和 MySQL 的 DDL 变更有所区别：
+
+| 语句类型 | 描述 | 错误信息示例 |
+| ------ | ---- | ------ | ------ |
+| Alter Table | 不支持在单个 alter 语句中指定多个 alter 选项 | Unsupported multi schema change |
+| Table Option | 表选项除了 `AUTO_INCREMENT`、`CHARACTER SET`、`COLLATE`、`COMMENT` 以外，其他所有的选项都会被忽略 | N/A |
```
Comment (AilinKid) [inline mysql-compatibility.md:69]: auto_id_cache、auto_random_base

Comment (tangenta) [reply mysql-compatibility.md:69]: MySQL doesn't support them, I think they can be put to `auto-increment.md` and `auto-random.md`.

Comment (AilinKid) [reply mysql-compatibility.md:69]: ok

File: auto-random.md [hunk 6]
```diff
@@ -1,6 +1,7 @@
 ---
 title: AUTO_RANDOM
 category: reference
+summary: 本文介绍了 TiDB 的列属性 auto_random。
```
Comment (TomShawn) [inline auto-random.md:4]:
```suggestion
summary: 本文介绍了 TiDB 的 `AUTO_RANDOM` 列属性 。
```

File: auto-increment.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+create table t (a int primary key AUTO_INCREMENT)
+```
+
+```sql
+insert into t values (), (), (), ()
+```
+
+即使存在正在执行并发写入的其他 TiDB 实例，或者当前实例剩余的缓存 ID 数量不够，都不会影响分配值的连续性。
+
+### _tidb_rowid 的关联性
```
Comment (TomShawn) [inline auto-increment.md:120]:
```suggestion
### `_tidb_rowid` 的关联性
```

File: auto-increment.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+> 在集群中有多个 TiDB 实例时，如果表结构中有自增 ID，建议不要混用显式插入和隐式分配（即自增列的缺省值和自定义值），否则可能会破坏隐式分配值的唯一性。
+
+例如在上述示例中，依次执行如下操作：
+
+1. 客户端向 B 插入一条将 `id` 设置为 2 的语句 `insert into t values (2, 1)`，并执行成功。
+2. 客户端向 A 发送 Insert 语句 `insert into t (c) (1)`，这条语句中没有指定 `id` 的值，所以会由 A 分配，当前 A 缓存了 [1, 30000] 这段 ID，可能会分配 2 为自增 ID 的值，并把本地计数器加 1。而此时数据库中已经存在 `id` 为 2 的数据，最终返回 `Duplicated Error` 错误。
+
+### 递增性保证
+
+`AUTO_INCREMENT` 列隐式分配值的递增性在含有多个 TiDB 实例的集群中无法得到保证。
```
Comment (kissmydb) [inline auto-increment.md:100]: 说明一下单个 tidb-server 是可以保证自增？

File: auto-increment.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+|           4 |    1 |
+|           5 |    2 |
+|           6 |    3 |
++-------------+------+
+3 rows in set (0.01 sec)
+```
+
+### 缓存大小控制
+
+TiDB 自增 ID 的缓存大小在早期版本中是对用户透明的。从 v3.1.2、v3.0.14 和 v4.0.rc.2 版本开始，TiDB 引入了 `AUTO_ID_CACHE` 表选项来允许用户自主设置自增 ID 分配缓存的大小。此外如果在 `INSERT` 语句中所需连续 ID 长度超过 `AUTO_ID_CACHE` 的长度时，TiDB 会适当调大缓存以便能够保证该语句的正常插入。
```
Comment (kissmydb) [inline auto-increment.md:147]: 可以举一个设置`AUTO_ID_CACHE `的例子么？

File: auto-increment.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+## 使用限制
+
+目前在 TiDB 中使用 `AUTO_INCREMENT` 有以下限制：
+
+- 必须定义在主键或者唯一索引的列上。
+- 只能定义在类型为整数、`FLOAT` 或 `DOUBLE` 的列上。
+- 不支持与列的默认值 `DEFAULT` 同时指定在同一列上。
+- 不支持使用 `ALTER TABLE` 来添加 `AUTO_INCREMENT` 属性。
+- 支持使用 `ALTER TABLE` 来移除 `AUTO_INCREMENT` 属性。但从 TiDB 2.1.18 和 3.0.4 版本开始，TiDB 通过 session 变量 `@@tidb_allow_remove_auto_inc` 控制是否允许通过 `ALTER TABLE MODIFY` 或 `ALTER TABLE CHANGE` 来移除列的 `AUTO_INCREMENT` 属性，默认是不允许移除。
```
Comment (kissmydb) [inline auto-increment.md:161]: 一旦移除，无法再加入，这个也写一下吧

Comment (tangenta) [reply auto-increment.md:161]: - 不支持使用 `ALTER TABLE` 来添加 `AUTO_INCREMENT` 属性。

上一条已经说明无法添加。

**Comment by TomShawn** [comment]
@tangenta 这个 PR 是否适用于 4.0 版本呢？

**Comment by tangenta** [comment]
@TomShawn 4.0 版本可能要修改一点内容，具体是 DDL 兼容性表格的这部分要改为：
```
| Add Index | 所有的索引选项都会被忽略 | N/A | 
```

**Comment by TomShawn** [comment]
@tangenta 好的，Special Week 之后请提 PR 修改 4.0 文档对应的部分，thanks!

File: mysql-compatibility.md [hunk 11]
```diff
@@ -42,57 +43,13 @@ TiDB 支持 MySQL 传输协议及其绝大多数的语法。这意味着您现
...
-+-------------+------+
-|           4 |    1 |
-|           5 |    2 |
-|           6 |    3 |
-+-------------+------+
-3 rows in set (0.01 sec)
-```
-
-TiDB 自增 ID 的缓存大小在早期版本中是对用户透明的。从 v3.1.2、v3.0.14 和 v4.0.rc.2 版本开始，TiDB 引入了 `AUTO_ID_CACHE` 表选项来允许用户自主设置自增 ID 分配缓存的大小。其中缓存大小可能会被自增列和 `_tidb_rowid` 共同消耗。此外如果在 `INSERT` 语句中所需连续 ID 长度超过 `AUTO_ID_CACHE` 的长度时，TiDB 会适当调大缓存以便能够保证该语句的正常插入。
+TiDB 中，对于自增列隐式分配的值，只保证单节点的自增和全局节点的唯一，不保证连续性。TiDB 目前采用批量分配 ID 的方式，如果在多台 TiDB 上同时插入数据，分配的自增 ID 会不连续。此外，不恰当地混用自定义值和缺省值可能会破坏数据唯一性。详情参见 [AUTO_INCREMENT](/auto-increment.md)。
```
Comment (yikeke) [inline mysql-compatibility.md:46]: When your are translating this PR, please note that most of the changes to `mysql-compatibility.md` are overwritten by https://github.com/pingcap/docs-cn/pull/3166/. @TomShawn

Comment (TomShawn) [reply mysql-compatibility.md:46]: OK, thanks!

**Comment by yikeke** [comment]
> @tangenta 好的，Special Week 之后请提 PR 修改 4.0 文档对应的部分，thanks!



> @TomShawn 4.0 版本可能要修改一点内容，具体是 DDL 兼容性表格的这部分要改为：
> 
> ```
> | Add Index | 所有的索引选项都会被忽略 | N/A | 
> ```

Inapplicable content deleted in https://github.com/pingcap/docs-cn/pull/3436

**Pull request merged**